### PR TITLE
Fortify - Add onDeployStop event for cancelling before confirming

### PR DESCRIPTION
**Pull request by DartRuffian**
**When merged this pull request will:**
- _Add a_ `ace_fortify_onDeployStop` _local event for when deploying is cancelled before confirming_
  - I.e. is called when right clicking after the menu appears
  - Is also called when `ace_fortify_deployCanceled` is

### IMPORTANT

- If the contribution affects [the documentation](https://github.com/acemod/ACE3/tree/master/docs), please include your changes in this pull request so the documentation will appear on the [website](https://ace3.acemod.org/).
- [Development Guidelines](https://ace3.acemod.org/wiki/development/) are read, understood and applied.
- Title of this PR uses our standard template `Component - Add|Fix|Improve|Change|Make|Remove {changes}`.

### CODE REVIEW

**Comment by DartRuffian** [comment]
Title could probably be changed, feels a bit wordy but couldn't think of a good way to phrase it.

**Comment by johnb432** [comment]
Why not use `ace_fortify_deployCanceled`, instead of creating a new event?

**Comment by DartRuffian** [comment]
> Why not use `ace_fortify_deployCanceled`, instead of creating a new event?

Different parameters, I made it match the `acex_fortify_onDeployStart` event, subbing out the object itself with the type of object.

**Comment by johnb432** [comment]
> Different parameters, I made it match the `acex_fortify_onDeployStart` event, subbing out the object itself with the type of object.

`deployCanceled` contains the same parameters (and more) as your current implementation of `ace_fortify_onDeployStop`, so imo not a good enough reason to add.

I'm guessing you want a listenable event when the player stops deploying?

**Comment by DartRuffian** [comment]
> > Different parameters, I made it match the `acex_fortify_onDeployStart` event, subbing out the object itself with the type of object.
> 
> `deployCanceled` contains the same parameters (and more) as your current implementation of `ace_fortify_onDeployStop`, so imo not a good enough reason to add.
> 
> I'm guessing you want a listenable event when the player stops deploying?

I could give it the same params, just figured it would make more sense to have a matching event instead of the `deployCanceled`. Could just update it to use the other event as well.

**Comment by johnb432** [comment]
> Updated it to use `deployCanceled` instead of a new event

You misunderstood me and my intentions.

`deployCanceled` does not work in this context, because we refund the cost of placing something down if you cancel it (see `XEH_postInit.sqf)`.

> deployCanceled contains the same parameters (and more) as your current implementation of ace_fortify_onDeployStop, so imo not a good enough reason to add.

There might be a good reason to add a new event, however the one you gave isn't good enough.

So I ask again: I'm guessing you want a listenable event when the player stops deploying?
If yes, then, given how the component is structured, we should add a listenable event.

**Comment by DartRuffian** [comment]
> There might be a good reason to add a new event, however the one you gave isn't good enough.
> 
> So I ask again: I'm guessing you want a listenable event when the player stops deploying? If yes, then, given how the component is structured, we should add a listenable event.

Yes.
Currently, `onDeployCanceled` is only emitted after the player initially starts building an object. If you have code that runs on `acex_fortify_onDeployStart` and is supposed to stop running if a player stops building an object, there is no (easy) way to detect if a player stops building before starting to construct an object.

**Comment by johnb432** [comment]
@DartRuffian I've added your initial suggestion back partially. However, this event won't be thrown when you cancel.
If you want to know when it's cancelled, use the cancel event. I see no reason why to mix stop and cancel in the stop event. It would make it worse imo: You couldn't tell the difference between when something has been cancelled or stopped, when there might be cases where you'd want to.

File: addons/fortify/functions/fnc_deployObject.sqf [hunk 1]
```diff
@@ -70,6 +70,7 @@ private _mouseClickID = [_player, "DefaultAction", {GVAR(isPlacing) == PLACE_WAI
             [_unit, _object] call FUNC(deployConfirm);
         } else {
             TRACE_1("deleting object",_object);
+            [QGVAR(onDeployStop), [_unit, _object, _cost]] call CBA_fnc_localEvent;
```
Comment (johnb432) [inline addons/fortify/functions/fnc_deployObject.sqf:73]: Not sure whether `typeOf _object` is more appropriate, given that _object is deleted in the next line.
On the other hand, it would allow 3rd party hooks to run code on object before it's deleted.

**Comment by DartRuffian** [comment]
> @DartRuffian I've added your initial suggestion back partially. However, this event won't be thrown when you cancel. If you want to know when it's cancelled, use the cancel event. I see no reason why to mix stop and cancel in the stop event. It would make it worse imo: You couldn't tell the difference between when something has been cancelled or stopped, when there might be cases where you'd want to.

I think I had both events originally there as a placeholder, can't remember an actual good reason for it being there

File: addons/fortify/functions/fnc_deployObject.sqf (hunk 1 above)
Comment (DartRuffian) [reply addons/fortify/functions/fnc_deployObject.sqf:73]: I feel like `typeOf _object` is more fitting because the object is being deleted on the next line anyway. It seems almost misleading to pass the object to an event handler, but delete it right after emitting the event.

Comment (johnb432) [reply addons/fortify/functions/fnc_deployObject.sqf:73]: We can delay the deleting of the object by a frame if necessary, I don't think it would pose too much of an issue.

The reason why `_object` is more interesting than `typeOf _object` is because the former can be derived easily from the prior and you might want to get e.g. `vectorUp` or other measurements related to the physical object's placement in the world.

@LinkIsGrim I need input on this.

Comment (LinkIsGrim) [reply addons/fortify/functions/fnc_deployObject.sqf:73]: I'm not a fan of delaying object deletion. All eventhandlers for `QGVAR(onDeployStop)` will execute before the object is deleted.

Comment (johnb432) [reply addons/fortify/functions/fnc_deployObject.sqf:73]: So we can leave it as is?

Comment (LinkIsGrim) [reply addons/fortify/functions/fnc_deployObject.sqf:73]: Yes

**Review by johnb432 — APPROVED** [review]
